### PR TITLE
Job new state machine

### PR DESCRIPTION
**Pull request by EvgenKor**
- rebuilt job state machine in library
- `waitJobCompleted` and `waitJobsCompleted` renamed to `waitJobTxHash` and `waitJobsTxHashes` accordingly
- introducing job monitoring promises
- new possible error: `PoolJobError`

This branch can be tested with zkbob-console branch [job-new-statemachine](https://github.com/zkBob/zkbob-console/tree/job-new-statemachine)

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -280,77 +286,158 @@ export class ZkBobClient {
...
   }
 
-  // Waiting while relayer process the job
-  // return transaction(s) hash(es) on success or throw an error
-  public async waitJobCompleted(tokenAddress: string, jobId: string): Promise<string[]> {
+  // Waiting while relayer process the job and send it to the Pool
+  // return transaction hash on success or throw an error
+  public async waitJobTxHash(tokenAddress: string, jobId: string): Promise<string> {
     const token = this.tokens[tokenAddress];
     const state = this.zpStates[tokenAddress];
```
[the remark this answers is not in the document]
Comment (EvgenKor) [reply src/client.ts]: These lines were removed

File: src/client.ts [hunk 2]
```diff
@@ -280,77 +286,158 @@ export class ZkBobClient {
...
-      } else if (job.state === 'failed') {
-        const relayerReason = job.failedReason !== undefined ? job.failedReason : 'unknown reason';
-        state.history.setQueuedTransactionFailedByRelayer(jobId, relayerReason);
-        throw new RelayerJobError(Number(jobId), job.failedReason !== undefined ? job.failedReason : 'unknown reason');
-      } else if (job.state === 'completed') {
-        hashes = job.txHash;
-        break;
+      } else {
+        job = jobInfo;
+        let jobDescr = `job #${jobId}${job.resolvedJobId != jobId ? `(->${job.resolvedJobId})` : ''}`;
```
[the remark this answers is not in the document]
Comment (EvgenKor) [reply src/client.ts]: Yes, agreed. Good reason to refactor code :) Done in https://github.com/zkBob/zkbob-client-js/pull/55/commits/65ce5c71ecaa1b93c7c708abb45146bd7a1eee0c

File: src/networks/evm.ts [hunk 3]
```diff
@@ -181,4 +181,29 @@ export class EvmNetwork implements NetworkBackend {
...
 
+    public async getTxRevertReason(txHash: string): Promise<string | null> {
+        const txReceipt = await this.web3.eth.getTransactionReceipt(txHash);
+        if (txReceipt && txReceipt.status !== undefined) {
+            if (txReceipt.status == false) {
+                const txData = await this.web3.eth.getTransaction(txHash);
+                
+                let reason = 'unknown reason';
+                try {
+                    await this.web3.eth.call(txData, txData.blockNumber)
```
[the remark this answers is not in the document]
Comment (EvgenKor) [reply src/networks/evm.ts]: I have fixed it in the last commit. But I can't get in mind why is it necessary? Could you explain pls?

[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply src/networks/evm.ts]: After this change `Web3` type can be restored:
https://github.com/zkBob/zkbob-client-js/blob/65ce5c71ecaa1b93c7c708abb45146bd7a1eee0c/src/networks/evm.ts#L11

Not critical, just to preserve type safety :)

[the remark this answers is not in the document]
Comment (EvgenKor) [reply src/networks/evm.ts]: Thanks, it's restored now in https://github.com/zkBob/zkbob-client-js/pull/55/commits/b09c61e322cd4faa79a46620809e02ad2b9332c0

**Review by AllFi — APPROVED** [review]
It works properly and looks ok to me.

**Review by k1rill-fedoseev — APPROVED** [review]

**Review by LyzeOfKiel — APPROVED** [review]

**Pull request merged**